gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -328,7 +328,9 @@ boolean shouldUseS2A() {
 
       // If a custom endpoint is being used, skip S2A.
       if (!Strings.isNullOrEmpty(clientSettingsEndpoint())
-          || !Strings.isNullOrEmpty(transportChannelProviderEndpoint())) {
+          || (!Strings.isNullOrEmpty(transportChannelProviderEndpoint())
+              && !buildEndpointTemplate(serviceName(), resolvedUniverseDomain())
+                  .contains(transportChannelProviderEndpoint()))) {
```

Comment:
Adding this "contains" logic, since `buildEndpointTemplate` adds `443` (https) as default port, and [`GrpcTransportOptions.setUpChannelProvider`](https://github.com/googleapis/sdk-platform-java/blob/2a41393c0e148e96fd3a0a8e1505e0827aa4fd3a/java-core/google-cloud-core-grpc/src/main/java/com/google/cloud/grpc/GrpcTransportOptions.java#L165), for example, just sets the host.

In this way, we are sure to respect a custom endpoint set via `TransportChannelProvider`.

@lqiu96

Reply:
> Could this check for both clientSettings and transportChannelProvider just to be safe? Or is there a specific reason for only doing this for the transportchannelprovider?

I don't see an issue for doing this check for both `clientSettingsEndpoint` and `transportChannelProviderEndpoint`, the reason I initially did it for the latter was because I ran into it in the GrpcTransportOptions case :). I tested this internally and verified that doing this check on both is ok. Added this in the latest commit.

Reply:
FYI, Bigtable has custom code to setEndpoint via settings which I think would be an issue: https://github.com/googleapis/java-bigtable/blob/8d3dca43224179829829bcf91972610c666b130b/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java#L807

May be an issue in the future as it auto sets a default. But you would prefer to handle those cases later on, then we can just proceed with this to unblock logging.